### PR TITLE
[FLINK-16245] Decoupling user classloader from context classloader.

### PR DESCRIPTION
**Pull request by AHeise**
## What is the purpose of the change

Decoupling user class loader from context classloader.

Thus, user classloader can be unloaded even though a reference on the context classloader outlives the user code.

## Brief change log

- Adding SafetyNetWrapperClassLoader
- Use it for FlinkUserCodeClassLoaders
- Make sure LocalExecutor closes user classloader

## Verifying this change

Added unit test.

## Does this pull request potentially affect one of the following parts:

  - Dependencies (does it add or upgrade a dependency): (yes / **no**)
  - The public API, i.e., is any changed class annotated with `@Public(Evolving)`: (yes / **no**)
  - The serializers: (yes / **no** / don't know)
  - The runtime per-record code paths (performance sensitive): (yes / **no** / don't know)
  - Anything that affects deployment or recovery: JobManager (and its components), Checkpointing, Kubernetes/Yarn/Mesos, ZooKeeper: (yes / **no** / don't know)
  - The S3 file system connector: (yes / **no** / don't know)

## Documentation

  - Does this pull request introduce a new feature? (yes / **no**)
  - If yes, how is the feature documented? (**not applicable** / docs / JavaDocs / not documented)

### CODE REVIEW

File: flink-runtime/src/main/java/org/apache/flink/runtime/execution/librarycache/CloseableClassLoader.java [hunk 1]
```diff
@@ -0,0 +1,36 @@
...
+ * <p>Unfortunately, ClassLoader is not an interface and does not implement {@link Closeable}, such that it is
+ * impossible to cleanly use interfaces to return a closeable ClassLoader. This class aims to provide at least some
+ * abstraction of the underlying implementation.
+ */
+public abstract class CloseableClassLoader extends ClassLoader implements Closeable {
+	public CloseableClassLoader(ClassLoader parent) {
+		super(parent);
+	}
+
+	public CloseableClassLoader() {
```
Comment (zentol) [inline flink-runtime/src/main/java/org/apache/flink/runtime/execution/librarycache/CloseableClassLoader.java]: is this constructor necessary?

File: flink-runtime/src/main/java/org/apache/flink/runtime/execution/librarycache/FlinkUserCodeClassLoaders.java [hunk 2]
```diff
@@ -82,4 +91,78 @@ public static ResolveOrder fromString(String resolveOrder) {
...
+				final Class<?> loadedClass = findLoadedClass(name);
+				if (loadedClass != null) {
+					return resolveIfNeeded(resolve, loadedClass);
+				}
+
+				return resolveIfNeeded(resolve, inner.loadClass(name));
+			}
+		}
+
+		private Class<?> resolveIfNeeded(final boolean resolve, final Class<?> loadedClass) {
```
Comment (zentol) [inline flink-runtime/src/main/java/org/apache/flink/runtime/execution/librarycache/FlinkUserCodeClassLoaders.java]: why is this method necessary? Shouldn't the inner classloader handle this?

File: flink-runtime/src/main/java/org/apache/flink/runtime/execution/librarycache/FlinkUserCodeClassLoaders.java [hunk 3]
```diff
@@ -82,4 +91,78 @@ public static ResolveOrder fromString(String resolveOrder) {
...
+			inner = null;
+		}
+
+		@Override
+		protected Class<?> loadClass(String name, boolean resolve) throws ClassNotFoundException {
+			if (inner == null) {
+				return super.loadClass(name, resolve);
+			}
+
+			synchronized (getClassLoadingLock(name)) {
```
Comment (zentol) [inline flink-runtime/src/main/java/org/apache/flink/runtime/execution/librarycache/FlinkUserCodeClassLoaders.java]: why aren't we just forwarding the call to the inner class?

File: flink-runtime/src/main/java/org/apache/flink/runtime/execution/librarycache/FlinkUserCodeClassLoaders.java (hunk 2 above)
Comment (AHeise) [reply flink-runtime/src/main/java/org/apache/flink/runtime/execution/librarycache/FlinkUserCodeClassLoaders.java]: `loadClass(String name, boolean resolve)` is protected, `loadClass(String name)` is not resolving by default.

File: flink-runtime/src/main/java/org/apache/flink/runtime/execution/librarycache/FlinkUserCodeClassLoaders.java (hunk 3 above)
Comment (AHeise) [reply flink-runtime/src/main/java/org/apache/flink/runtime/execution/librarycache/FlinkUserCodeClassLoaders.java]: `loadClass(String name, boolean resolve)` is protected, `loadClass(String name)` is not resolving by default.

Comment (zentol) [reply flink-runtime/src/main/java/org/apache/flink/runtime/execution/librarycache/FlinkUserCodeClassLoaders.java]: Since we only use it for our own custom classloaders we could modify the access level

Comment (AHeise) [reply flink-runtime/src/main/java/org/apache/flink/runtime/execution/librarycache/FlinkUserCodeClassLoaders.java]: `inner` is typically a `URLClassLoader`. We could subclass it, but I'd actually like to avoid that unless you have a strong opinion on that.

File: flink-runtime/src/main/java/org/apache/flink/runtime/execution/librarycache/FlinkUserCodeClassLoaders.java [hunk 4]
```diff
@@ -18,35 +18,44 @@
...
+import java.util.Enumeration;
 
 /**
  * Gives the URLClassLoader a nicer name for debugging purposes.
  */
 public class FlinkUserCodeClassLoaders {
 
-	public static URLClassLoader parentFirst(URL[] urls, ClassLoader parent) {
+	@VisibleForTesting
+	static URLClassLoader parentFirst(URL[] urls, ClassLoader parent) {
```
Comment (StephanEwen) [inline flink-runtime/src/main/java/org/apache/flink/runtime/execution/librarycache/FlinkUserCodeClassLoaders.java]: I would not reduce the visibility here, that is unrelated to the change, and I actually know users that have used this,

File: flink-runtime/src/main/java/org/apache/flink/runtime/execution/librarycache/FlinkUserCodeClassLoaders.java (hunk 3 above)
Comment (StephanEwen) [reply flink-runtime/src/main/java/org/apache/flink/runtime/execution/librarycache/FlinkUserCodeClassLoaders.java]: Given that `inner` is always a `Flink[Parent|Child]FirstClassLoader`, we could increase the visibility of the method.

File: flink-runtime/src/main/java/org/apache/flink/runtime/execution/librarycache/FlinkUserCodeClassLoaders.java (hunk 4 above)
Comment (AHeise) [reply flink-runtime/src/main/java/org/apache/flink/runtime/execution/librarycache/FlinkUserCodeClassLoaders.java]: Hm it's not public API and they could achieve the same thing through `#create` with one extra parameter. 
Unless I change the signature of this method, this method cannot use the safety net (or safety net would also need to be URLClassLoader which I wanted to avoid), which may also be fine, but I wanted to point that out explicitly.

File: flink-runtime/src/main/java/org/apache/flink/runtime/execution/librarycache/FlinkUserCodeClassLoaders.java (hunk 3 above)
Comment (AHeise) [reply flink-runtime/src/main/java/org/apache/flink/runtime/execution/librarycache/FlinkUserCodeClassLoaders.java]: We would need to introduce a common ancestor to `Flink[Parent|Child]FirstClassLoader` that does nothing but exposes `loadClass`. 
I'm really not convinced that this is better than do the resolution as is. It would certainly not be more concise and imho it would increase the complexity.

File: flink-runtime/src/main/java/org/apache/flink/runtime/execution/librarycache/FlinkUserCodeClassLoaders.java (hunk 4 above)
Comment (zentol) [reply flink-runtime/src/main/java/org/apache/flink/runtime/execution/librarycache/FlinkUserCodeClassLoaders.java]: Why do you not want the safety net to be a URLClassLoader?

Comment (AHeise) [reply flink-runtime/src/main/java/org/apache/flink/runtime/execution/librarycache/FlinkUserCodeClassLoaders.java]: I was not considering it initially because delegation pattern on implementation level felt wrong, but it's messy for ClassLoader anyways and it would solve this discussion, so I adjusted the PR.
Now we just need to converge on `loadClass`.

Comment (StephanEwen) [reply flink-runtime/src/main/java/org/apache/flink/runtime/execution/librarycache/FlinkUserCodeClassLoaders.java]: Fair enough

File: flink-runtime/src/main/java/org/apache/flink/runtime/execution/librarycache/FlinkUserCodeClassLoaders.java (hunk 3 above)
Comment (StephanEwen) [reply flink-runtime/src/main/java/org/apache/flink/runtime/execution/librarycache/FlinkUserCodeClassLoaders.java]: My guess is that there will be a common ancestor anyways int he future (to register disposal hooks, see [here](https://github.com/apache/flink/pull/11196#issuecomment-591347938)). But I am okay with having the in a follow-up effort.

File: flink-runtime/src/main/java/org/apache/flink/runtime/execution/librarycache/FlinkUserCodeClassLoaders.java (hunk 2 above)
Comment (zentol) [reply flink-runtime/src/main/java/org/apache/flink/runtime/execution/librarycache/FlinkUserCodeClassLoaders.java]: Now that the safety net is a URLClassLoader we can just forward the call, correct?

**Comment by AHeise** [comment]
@flinkbot run travis

**Comment by AHeise** [comment]
@flinkbot run azure

File: flink-runtime/src/main/java/org/apache/flink/runtime/execution/librarycache/FlinkUserCodeClassLoaders.java (hunk 2 above)
Comment (AHeise) [reply flink-runtime/src/main/java/org/apache/flink/runtime/execution/librarycache/FlinkUserCodeClassLoaders.java]: No, still not possible.

Comment (AHeise) [reply flink-runtime/src/main/java/org/apache/flink/runtime/execution/librarycache/FlinkUserCodeClassLoaders.java]: Added a common ancestor now.

**Comment by AHeise** [comment]
@flinkbot run azure

File: flink-runtime/src/main/java/org/apache/flink/runtime/execution/librarycache/FlinkUserCodeClassLoaders.java (hunk 3 above)
Comment (AHeise) [reply flink-runtime/src/main/java/org/apache/flink/runtime/execution/librarycache/FlinkUserCodeClassLoaders.java]: Added a common ancestor now and used delegation.

File: flink-runtime/src/main/java/org/apache/flink/runtime/execution/librarycache/FlinkUserCodeClassLoaders.java [hunk 5]
```diff
@@ -82,4 +89,61 @@ public static ResolveOrder fromString(String resolveOrder) {
...
+	 *
+	 * <p>This classloader delegates to the actual user classloader. Upon {@link #close()}, the delegate is nulled
+	 * and can be garbage collected. Additional class resolution will be resolved solely through the bootstrap
+	 * classloader and most likely result in ClassNotFound exceptions.
+	 */
+	private static class SafetyNetWrapperClassLoader extends URLClassLoader
+			implements Closeable {
+		private static final Logger LOG = LoggerFactory.getLogger(SafetyNetWrapperClassLoader.class);
+
+		private FlinkUserCodeClassLoader inner;
```
Comment (zentol) [inline flink-runtime/src/main/java/org/apache/flink/runtime/execution/librarycache/FlinkUserCodeClassLoaders.java]:
```suggestion
		private final FlinkUserCodeClassLoader inner;
```

**Comment by AHeise** [comment]
@flinkbot run azure

**Comment by AHeise** [comment]
@flinkbot run azure

**Comment by AHeise** [comment]
@flinkbot run azure

File: flink-tests/src/test/java/org/apache/flink/test/recovery/BatchFineGrainedRecoveryITCase.java [hunk 6]
```diff
@@ -396,14 +397,16 @@ private TaskExecutorFailureStrategy(int failAfterCallNumber) {
...
 			//noinspection OverlyBroadCatchBlock
-			try {
-				restartTaskManager();
-			} catch (InterruptedException e) {
-				// ignore the exception, task should have been failed while stopping TM
-				Thread.currentThread().interrupt();
-			} catch (Throwable t) {
-				failureTracker.unrelatedFailure(t);
-				throw t;
+			try (TemporaryClassLoaderContext unused = TemporaryClassLoaderContext.of(ClassLoader.getSystemClassLoader())) {
```
Comment (zentol) [inline flink-tests/src/test/java/org/apache/flink/test/recovery/BatchFineGrainedRecoveryITCase.java]: why do we need a separate classloader here?

File: flink-table/flink-sql-client/src/main/java/org/apache/flink/table/client/gateway/local/LocalExecutor.java [hunk 7]
```diff
@@ -240,7 +240,11 @@ public void closeSession(String sessionId) throws SqlExecutionException {
 			}
 		});
 		// Remove the session's ExecutionContext from contextMap.
-		this.contextMap.remove(sessionId);
+		try {
+			this.contextMap.remove(sessionId).close();
+		} catch (IOException e) {
+			// ignore any throwable to keep the clean up running
```
Comment (zentol) [inline flink-table/flink-sql-client/src/main/java/org/apache/flink/table/client/gateway/local/LocalExecutor.java]: should still be logged

File: flink-runtime/src/main/java/org/apache/flink/runtime/execution/librarycache/FlinkUserCodeClassLoaders.java [hunk 8]
```diff
@@ -82,4 +89,61 @@ public static ResolveOrder fromString(String resolveOrder) {
...
+					LOG.warn("Could not close user classloader", e);
+				}
+			}
+			inner = null;
+		}
+
+		@Override
+		protected Class<?> loadClass(String name, boolean resolve) throws ClassNotFoundException {
+			if (inner == null) {
+				return super.loadClass(name, resolve);
```
Comment (zentol) [inline flink-runtime/src/main/java/org/apache/flink/runtime/execution/librarycache/FlinkUserCodeClassLoaders.java]: Why are we not failing immediately in this case? This introduces subtle inconsistent behaviors, that to fully grasp require a good understanding of how class-loading works in the JVM which really shows in the added `testClosingOfClassloader` test. I would wager that barely anyone would be able to explain why the class cannot be loaded at the end of the test.

File: flink-tests/src/test/java/org/apache/flink/test/recovery/BatchFineGrainedRecoveryITCase.java (hunk 6 above)
Comment (AHeise) [reply flink-tests/src/test/java/org/apache/flink/test/recovery/BatchFineGrainedRecoveryITCase.java]: The test is badly written, but I haven't found a better solution. We are restarting a task manager from within a UDF (with context classloader set to user class loader). Restarting means closing TM, which closes context class loader, such that starting the TM fails.

File: flink-table/flink-sql-client/src/main/java/org/apache/flink/table/client/gateway/local/LocalExecutor.java (hunk 7 above)
Comment (AHeise) [reply flink-table/flink-sql-client/src/main/java/org/apache/flink/table/client/gateway/local/LocalExecutor.java]: Should we then also log on cancelQuery on L237? I tried to keep it symmetric.

File: flink-runtime/src/main/java/org/apache/flink/runtime/execution/librarycache/FlinkUserCodeClassLoaders.java (hunk 8 above)
Comment (AHeise) [reply flink-runtime/src/main/java/org/apache/flink/runtime/execution/librarycache/FlinkUserCodeClassLoaders.java]: I understand where you are coming from, but this will effectively make any job with leaked context classloader fail eventually, even if just basic Java stuff needs to be loaded. 
If we want to go that route, we should think twice what to throw. An `IllegalStateException` would be the correct one, but does it fit in all cases? We could also throw a `ClassNotFoundException` to not change the behavior too much and provide a good description, but that may trigger some unwanted fallback logic in user libraries.

Comment (zentol) [reply flink-runtime/src/main/java/org/apache/flink/runtime/execution/librarycache/FlinkUserCodeClassLoaders.java]: > any job with leaked context classloader [...] fail eventually

Given that the CL is only closed when all tasks on the TE have stopped, how could it affect a _job_? Leftover threads, sure, but what happens to those shouldn't affect job execution, and failing as early as possible might actually be a benefit.

> [throwing a CNFE] may trigger some unwanted fallback logic in user libraries

This can happen in any case; for all we now the library could be catching throwables.

I'd fail with an IllegalStateException; I think when users see a CNFE they first consider it as a class-loading issue in Flink (regardless of the error message).

Comment (AHeise) [reply flink-runtime/src/main/java/org/apache/flink/runtime/execution/librarycache/FlinkUserCodeClassLoaders.java]: > Given that the CL is only closed when all tasks on the TE have stopped, how could it affect a job? Leftover threads, sure, but what happens to those shouldn't affect job execution, and failing as early as possible might actually be a benefit.

I actually meant it the other way around. What happens if we have job1 and job2, where the CL of job1 is leaked forward while executing job2. The TM would die taking job2 with it with a somewhat unrelated error. In particular, I'd think it would be pretty confusing/inperformant for larger batch clusters. @StephanEwen WDYT?

But since I'm also a fan of failing early, I'd go that way once Stephan agrees (I kinda had the impression that he wanted it to be graceful).

Comment (zentol) [reply flink-runtime/src/main/java/org/apache/flink/runtime/execution/librarycache/FlinkUserCodeClassLoaders.java]: Why would the TM die because of an exception in user-code, disconnected from any task threads no less?

Comment (AHeise) [reply flink-runtime/src/main/java/org/apache/flink/runtime/execution/librarycache/FlinkUserCodeClassLoaders.java]: I thought that's what you meant with fail early. If you look into https://issues.apache.org/jira/browse/FLINK-11205, you can see the main motivation for this PR.

Here Stephan wrote 

> Some thoughts on how to approach this are:
> ...
> As a final safety-net, the TMs kill/restart themselves when the metaspace blows up FLINK-16225

Btw, this answers your initial point in this thread: throwing ISE it is. Now, I just need to figure out how to make sure that this exception will actually let TM fail instead of any later job.

Comment (zentol) [reply flink-runtime/src/main/java/org/apache/flink/runtime/execution/librarycache/FlinkUserCodeClassLoaders.java]: > I thought that's what you meant with fail early.

No, with fail early I meant that the (leaked) thread with the leaked classloader reference would crash ASAP if it tries to load a class when the CL was closed.

> Now, I just need to figure out how to make sure that this exception will actually let TM fail.

I don't see why we should do this.

FLINK-16225 aims to have the TM kill itself on a Metaspace OOM which is fine, but I find no reference anywhere that we should always fail the TM at the slightest hint of a leaked thread and/or classloader, which in any case would be out-of-scope of this PR?

Comment (AHeise) [reply flink-runtime/src/main/java/org/apache/flink/runtime/execution/librarycache/FlinkUserCodeClassLoaders.java]: > No, with fail early I meant that the (leaked) thread with the leaked classloader reference would crash ASAP if it tries to load a class when the CL was closed.

That sounds completely fine with me, but are we sure the CL is only leaked by leaking the thread? I understood that the Apache Commons Logging framework lives outside of the thread (in app classloader) and keeps the reference to the user classloader for the whole TM lifecycle.

Comment (zentol) [reply flink-runtime/src/main/java/org/apache/flink/runtime/execution/librarycache/FlinkUserCodeClassLoaders.java]: I think we can be sure a class loaded from the user-jar can only leak something through a separate thread.

But we aren't sure at all what libraries in /lib are doing.
JCL for example does keep references around (and even [documents](https://cwiki.apache.org/confluence/display/COMMONS/Logging+UndeployMemoryLeak) that), but I don't see a way to handle this from our end.

My conclusions would be that users should stop throwing dependencies into /lib and create fatter user-jars instead.

Comment (StephanEwen) [reply flink-runtime/src/main/java/org/apache/flink/runtime/execution/librarycache/FlinkUserCodeClassLoaders.java]: My feeling is to go with Arvid's solution. Keep the loading of classes that are not from the user code classloader intact.

File: flink-table/flink-sql-client/src/main/java/org/apache/flink/table/client/gateway/local/LocalExecutor.java (hunk 7 above)
Comment (zentol) [reply flink-table/flink-sql-client/src/main/java/org/apache/flink/table/client/gateway/local/LocalExecutor.java]: I think so, as a separate commit of course ;)

File: flink-runtime/src/main/java/org/apache/flink/runtime/execution/librarycache/FlinkUserCodeClassLoaders.java (hunk 8 above)
Comment (zentol) [reply flink-runtime/src/main/java/org/apache/flink/runtime/execution/librarycache/FlinkUserCodeClassLoaders.java]: Are we hoping that these threads shutdown gracefully if we allow them to load java classes, or are we worried that crashing them this way can cause side-effects?

Comment (StephanEwen) [reply flink-runtime/src/main/java/org/apache/flink/runtime/execution/librarycache/FlinkUserCodeClassLoaders.java]: My thought would be mainly understandable behavior. The change targets cases of caching beyond execution of slot related tasks, such as in logging frameworks. Users would really wonder why java core classes could not be loaded any more.

Aside from that, if the logging framwork only accesses core classes, this change would break nothing, while forbidding to load all classes might break some setups.

Related: Might make sense to wrap the `ClassNotFoundException` and enhance the error message with "Flink user code classloader is shut down".

Comment (zentol) [reply flink-runtime/src/main/java/org/apache/flink/runtime/execution/librarycache/FlinkUserCodeClassLoaders.java]: > Users would really wonder why java core classes could not be loaded any more.

That's a fair point.

**Comment by zentol** [comment]
@flinkbot run azure
